src/reflect/scala/reflect/internal/Types.scala

```diff
@@ -2408,7 +2408,7 @@ trait Types
     override final def equals(other: Any): Boolean = {
       if (this eq other.asInstanceOf[AnyRef]) true
       else other match {
-        case otherTypeRef: TypeRef =>
+        case otherTypeRef: TypeRef if !(otherTypeRef.isInstanceOf[ObjectTpeJavaRef] || this.isInstanceOf[ObjectTpeJavaRef]) =>
           Objects.equals(pre, otherTypeRef.pre) && sym.eq(otherTypeRef.sym) && sameElementsEquals(args, otherTypeRef.args)
```

Comment:
We could put the new conditions after the existing ones if we're paranoid about performance.

```
 case otherTypeRef: TypeRef =>
  Objects.equals(pre, otherTypeRef.pre) && sym.eq(otherTypeRef.sym) && sameElementsEquals(args, otherTypeRef.args) && !(otherTypeRef.isInstanceOf[ObjectTpeJavaRef] || this.isInstanceOf[ObjectTpeJavaRef])
```